src/limitedmap.h

```diff
@@ -26,68 +26,76 @@ class limitedmap
+            if (x.second <  cur_min) {
+                return;
+            }
+            map.erase(rmap.begin()->second);
+            rmap.erase(rmap.begin());
+        }
         std::pair<iterator, bool> ret = map.insert(x);
         if (ret.second) {
-            if (nMaxSize && map.size() == nMaxSize) {
-                map.erase(rmap.begin()->second);
-                rmap.erase(rmap.begin());
-            }
-            rmap.insert(make_pair(x.second, ret.first));
+            // in zend, we expect to insert increasing timestamps most of the times, hence the hint.
+            rmap.insert(rmap.end(), make_pair(x.second, ret.first));
         }
-        return;
     }
-    size_t erase(const key_type& k)
-    {
+
+    size_t erase(const key_type& k) {
         iterator itTarget = map.find(k);
         if (itTarget == map.end())
             return 0;
-        std::pair<rmap_iterator, rmap_iterator> itPair = rmap.equal_range(itTarget->second);
-        for (rmap_iterator it = itPair.first; it != itPair.second; ++it) {
-            if (it->second == itTarget) {
-                rmap.erase(it);
-                map.erase(itTarget);
-                return 1; // be consistent with std::map return value
-            }
-        }
-        // Shouldn't ever get here
-        assert(0);
-        return 0;
+        rmap_iterator rit = find_rmap_key_value_pair(itTarget);
+        assert(rit != rmap.end());
+        rmap.erase(rit);
+        map.erase(itTarget);
+        return 1; // be consistent with std::map return value
     }
-    void update(const_iterator itIn, const mapped_type& v)
-    {
-        // TODO: When we switch to C++11, use map.erase(itIn, itIn) to get the non-const iterator.
-        iterator itTarget = map.find(itIn->first);
+
+    void update(const_iterator itIn, const mapped_type& v) {
+        iterator itTarget = map.erase(itIn, itIn); // empty range used to get iterator from const_iterator
         if (itTarget == map.end())
             return;
-        std::pair<rmap_iterator, rmap_iterator> itPair = rmap.equal_range(itTarget->second);
-        for (rmap_iterator it = itPair.first; it != itPair.second; ++it)
-            if (it->second == itTarget) {
-                rmap.erase(it);
-                itTarget->second = v;
-                rmap.insert(make_pair(v, itTarget));
-                return;
-            }
-        // Shouldn't ever get here
-        assert(0);
+
+        rmap_iterator rit = find_rmap_key_value_pair(itIn);
+        assert(rit != rmap.end());
+        rmap.erase(rit);
+        itTarget->second = v;
+        rmap.insert(std::make_pair(v, itTarget));
     }
+
     size_type max_size() const { return nMaxSize; }
-    size_type max_size(size_type s)
-    {
-        if (s)
-            while (map.size() > s) {
-                map.erase(rmap.begin()->second);
-                rmap.erase(rmap.begin());
-            }
+
+    size_type max_size(size_type s) {
```

Comment:
As for the ctor I'd drop an assertion for `s < 1`

Reply:
Yep